examples/controlnet/train_controlnet_flux.py

```diff
@@ -0,0 +1,1322 @@
+        args.pretrained_model_name_or_path,
+        subfolder="vae",
+        revision=args.revision,
+        variant=args.variant,
+    )
+    flux_transformer = FluxTransformer2DModel.from_pretrained(
+        args.pretrained_model_name_or_path,
+        subfolder="transformer",
+        revision=args.revision,
+        variant=args.variant,
+    )
+    if args.flux_controlnet_model_name_or_path:
+        logger.info("Loading existing controlnet weights")
+        flux_controlnet = FluxControlNetModel.from_pretrained(args.flux_controlnet_model_name_or_path)
+    else:
+        logger.info("Initializing controlnet weights from transformer")
+        # we can define the num_layers, num_single_layers,
+        # default values are num_layers=4, num_single_layers=10
+        flux_controlnet = FluxControlNetModel.from_transformer(
+            flux_transformer,
+            num_layers=args.num_double_layers,
+            num_single_layers=args.num_single_layers,
+        )
+    logger.info("all models loaded successfully")
+
+    noise_scheduler = FlowMatchEulerDiscreteScheduler.from_pretrained(
+        args.pretrained_model_name_or_path,
+        subfolder="scheduler",
+    )
+
+    vae.requires_grad_(False)
+    flux_transformer.requires_grad_(False)
+    text_encoder_one.requires_grad_(False)
+    text_encoder_two.requires_grad_(False)
+    flux_controlnet.train()
+
+    # use some pipeline function
+    flux_controlnet_pipeline = FluxControlNetPipeline(
+        scheduler=noise_scheduler,
+        vae=vae,
+        text_encoder=text_encoder_one,
+        tokenizer=tokenizer_one,
+        text_encoder_2=text_encoder_two,
+        tokenizer_2=tokenizer_two,
+        transformer=flux_transformer,
+        controlnet=flux_controlnet,
+    ).to(accelerator.device)
+
+    def unwrap_model(model):
+        model = accelerator.unwrap_model(model)
+        model = model._orig_mod if is_compiled_module(model) else model
+        return model
+
+    # `accelerate` 0.16.0 will have better support for customized saving
+    if version.parse(accelerate.__version__) >= version.parse("0.16.0"):
+        # create custom saving & loading hooks so that `accelerator.save_state(...)` serializes in a nice format
+        def save_model_hook(models, weights, output_dir):
+            if accelerator.is_main_process:
+                i = len(weights) - 1
+
+                while len(weights) > 0:
+                    weights.pop()
+                    model = models[i]
+
+                    sub_dir = "flux_controlnet"
+                    model.save_pretrained(os.path.join(output_dir, sub_dir))
+
+                    i -= 1
+
+        def load_model_hook(models, input_dir):
+            while len(models) > 0:
+                # pop models so that they are not loaded again
+                model = models.pop()
+
+                # load diffusers style into model
+                load_model = FluxControlNetModel.from_pretrained(input_dir, subfolder="controlnet")
```

Comment:
Should `subfolder` here be the same name used for `sub_dir` inside `save_model_hook`? Seems they are different.

Reply:
Thanks, fixed